### PR TITLE
Implement hpx::experimental::for_each_index (P4150R0)

### LINKED ISSUE #7214: Implement `hpx::experimental::for_each_index` as described in P4150
see https://wg21.link/p4150

### PR DESCRIPTION
**Pull request by arpittkhandelwal**
Fixes #7214

# Description

This PR introduces `hpx::experimental::for_each_index`, an implementation of the multidimensional index-based for-each algorithm proposed in ISO C++ proposal [P4150R0](https://wg21.link/p4150).

The algorithm iterates over the domain defined by the extents of a layout mapping and calls a function object with exactly `Mapping::rank()` index arguments, parallelizing the outermost extent and executing the inner extents sequentially.

### Features
* **Concept Guard**: Employs a lightweight, C++20-compatible `concept` (`is_layout_mapping_v`) to detect valid mapping types without relying on C++23 `<mdspan>`.
* **Sequential Optimizations**: Includes a compile-time variadic recursion algorithm `hpx::parallel::detail::for_each_index_seq_impl` to flatten multidimensional loop unrolling seamlessly.
* **Parallel Performance**: Parallel overloads leverage `hpx::parallel::util::partitioner` over the rank-0 extent, while preserving data-locality for row-major layouts via local inner sequential steps.
* **Execution Policies Supported**: Operates seamlessly with all established HPX execution policy behaviors, including `seq`, `par`, `par_unseq`, and `task` variations.
* **Comprehensive Test Suite**: Incorporates a lightweight `simple_mapping` stub for ensuring tests accurately mimic a full multidimensional array without demanding ISO C++23 capability.

### Changed Files
- `libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp` (New)
- `libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp` (New)
- `libs/core/algorithms/include/hpx/parallel/algorithm.hpp`
- `libs/core/algorithms/CMakeLists.txt`
- `libs/core/algorithms/tests/unit/algorithms/CMakeLists.txt`

Tested formatting via clang-format locally against the HPX format spec.

### CODE REVIEW

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+      : ext_(dims)
+    {
+    }
+
+    // rank-0 specialisation helper – constructed with no dims.
+    simple_mapping() = default;
+
+    [[nodiscard]] static constexpr std::size_t rank() noexcept
+    {
+        return Rank;
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp]: Fixed.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        // A type M satisfies the concept when it exposes:
+        //   - M::index_type   (member type)
+        //   - M::extents_type (member type)
+        //   - M::rank()       (static constexpr function returning std::size_t)
+        //   - m.extents()     (const member function)
+        template <typename M>
+        struct is_layout_mapping_impl<M,
+            std::void_t<typename M::index_type, typename M::extents_type,
+                decltype(M::rank()),
+                decltype(std::declval<M const&>().extents())>> : std::true_type
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Fixed.

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+///////////////////////////////////////////////////////////////////////////////
+int hpx_main(hpx::program_options::variables_map& vm)
+{
+    unsigned int seed = static_cast<unsigned int>(std::time(nullptr));
+    if (vm.count("seed"))
+        seed = vm["seed"].as<unsigned int>();
+
+    std::cout << "using seed: " << seed << std::endl;
+    std::srand(seed);
+
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp]: Fixed.

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+//
+// A minimal layout-mapping stub is used so the tests compile on any C++
+// standard mode without requiring C++23 <mdspan>.
+
+#include <hpx/algorithm.hpp>
+#include <hpx/init.hpp>
+#include <hpx/modules/testing.hpp>
+
+#include <atomic>
+#include <cstddef>
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp:17]: Fixed.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        HPX_HOST_DEVICE HPX_FORCEINLINE constexpr void operator()(
+            Iter part_begin, std::size_t part_size) const
+        {
+            for (std::size_t k = 0; k < part_size; ++k, ++part_begin)
+            {
+                auto i0 = static_cast<index_type>(*part_begin);
+                // Start recursion from dimension 1; pass i0 as the first
+                // collected index.  When rank==1 this immediately invokes fun.
+                for_each_index_seq_impl<1>(mapping_, fun_, i0);
+            }
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:241]: Fixed.

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp:17]: Please do not use `std::srand`, use `<random>` facilities instead (`std::random_device` and `std::mt19937`, see other tests for examples).

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp]: We usually let the tests generate random numbers for the inputs to trigger boundary cases.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            constexpr bool has_scheduler_executor =
+                hpx::execution_policy_has_scheduler_executor_v<ExPolicy>;
+
+            // rank-0: exactly one invocation – execute sequentially.
+            if constexpr (Mapping::rank() == 0)
+            {
+                HPX_INVOKE(f);
+                return util::detail::algorithm_result<ExPolicy>::get();
+            }
+            else
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:350]: This parallelizes things for the outermost rank only. Is that how it is specified in the paper?

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+    hpx::parallel::util::detail::algorithm_result_t<ExPolicy> for_each_index(
+        ExPolicy&& policy, Mapping const& mapping, Fun fun);
+
+}}    // namespace hpx::experimental
+
+#else
+
+#include <hpx/config.hpp>
+#include <hpx/functional/invoke.hpp>
+#include <hpx/iterator_support/counting_iterator.hpp>
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Please use the generated module headers:
```suggestion
#include <hpx/modules/functional.hpp>
#include <hpx/modules/iterator_support.hpp>
```

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+        if constexpr (RankIdx == Mapping::rank())
+        {
+            HPX_INVOKE(f, idx...);
+        }
+        else
+        {
+            using index_type = typename Mapping::index_type;
+            index_type const bound =
+                static_cast<index_type>(m.extents().extent(RankIdx));
+            for (index_type i = index_type{0}; i < bound; ++i)
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: If you used the appropriate overload from the `loop_n` facilities, your implementation would potentially support vectorization as well - at least for the innermost rank iteration.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+            for_each_index_iteration&&) = default;
+        ~for_each_index_iteration() = default;
+
+        // Called by the partitioner with [part_begin, part_begin + part_size)
+        // – a contiguous range of integer i0 values to process on this thread.
+        template <typename Iter>
+        HPX_HOST_DEVICE HPX_FORCEINLINE constexpr void operator()(
+            Iter part_begin, std::size_t part_size)
+        {
+            for (std::size_t k = 0; k < part_size; ++k, ++part_begin)
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Same here, use one of the existing `loop_n` overloads to potentially support vectorization.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                    for_each_index_iteration<Mapping, std::decay_t<Fun>>{
+                        m, HPX_FORWARD(Fun, f)};
+
+                if constexpr (hpx::is_async_execution_policy_v<ExPolicy> ||
+                    has_scheduler_executor)
+                {
+                    return util::detail::algorithm_result<ExPolicy>::get(
+                        util::partitioner<ExPolicy>::call(
+                            HPX_FORWARD(ExPolicy, policy), first, count,
+                            HPX_MOVE(iter_fun), hpx::util::empty_function{}));
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:434]: I believe that `algorithm_result<ExPolicy>::get` will work for all execution policies. There might be no need to separate the sync and the async case here.

**Review by hkaiser — COMMENTED** [review]
Thanks for looking into this! Great job!

**Comment by hkaiser** [comment]
Also, inpect reports an issue:
```
/libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:
    Non-ASCII: (line 175)
```

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp:17]: Fixed! I have replaced the std::srand and std::time setup with std::random_device and std::mt19937. The tests now use std::uniform_int_distribution to generate random dimensions while still supporting the --seed flag for reproducibility.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 7 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Done. I've updated the includes to use the module umbrella headers (<hpx/modules/functional.hpp>, <hpx/modules/iterator_support.hpp>, etc.).

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 8 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: I have refactored both the internal recursion (for_each_index_seq_impl) and the chunked operator in for_each_index_iteration to use hpx::parallel::util::loop_n with the provided execution policy. This ensures the algorithm can now leverage SIMD vectorization (e.g., with unseq or par_unseq) across all dimensions.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 10 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:434]: I've simplified the logic by removing the redundant if constexpr branches. We now unconditionally pass the partitioner result to util::detail::algorithm_result<ExPolicy>::get(), which handles the sync/async dispatch internally.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 6 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:350]: Yes, the paper (P4150R0, Sections 4.1 and 6) explicitly leaves the iteration order and parallelization strategy as an implementation-defined "hint."

Since mdspan defaults to row-major (layout_right), parallelizing the outermost rank and iterating the inner ranks sequentially provides the best spatial locality. It avoids cache thrashing when multiple threads access contiguous memory and fits perfectly with HPX's existing partitioner design.

**Comment by arpittkhandelwal** [comment]
> Also, inpect reports an issue:
> 
> ```
> /libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:
>     Non-ASCII: (line 175)
> ```

Fixed. I caught some en-dashes (–) in the comments that the inspect tool flagged. I've scrubbed the file and replaced them with standard ASCII hyphens.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 9 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Fixed

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    ///        \a mapping (i.e., its extents), executed according to \a policy.
+    ///
+    /// \note Complexity: Applies \a fun exactly the product of all extents of
+    ///       \a mapping times.
+    ///
+    /// If \a fun returns a result, the result is ignored.
+    ///
+    /// Unlike its sequential form, the parallel overload does not return a
+    /// copy of its \a Fun parameter.
+    ///
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: That was a copy-paste error from std::for_each's documentation. Since both the sequential and parallel overloads of for_each_index return void, the comparison was indeed misleading. I have removed that note from the docstring."

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+                iter_t first{index_type{0}};
+
+                auto iter_fun = for_each_index_iteration<std::decay_t<ExPolicy>,
+                    Mapping, std::decay_t<Fun>>{m, HPX_FORWARD(Fun, f)};
+
+                return util::detail::algorithm_result<ExPolicy>::get(
+                    util::partitioner<ExPolicy>::call(
+                        HPX_FORWARD(ExPolicy, policy), first, count,
+                        HPX_MOVE(iter_fun), hpx::util::empty_function{}));
+            }
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: I had simplified this based on an earlier suggestion, but you're absolutely right—passing a void expression into algorithm_result::get for synchronous policies is ill-formed and would break the build. I've reverted this to use the if constexpr pattern (matching for_loop) to correctly handle the synchronous and asynchronous/task-based paths separately.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+    // thread-safe and can be copied to GPU device memory if needed.
+    template <typename ExPolicy, typename Mapping, typename Fun>
+    struct for_each_index_iteration
+    {
+        using fun_type = std::decay_t<Fun>;
+        using index_type = typename Mapping::index_type;
+
+        Mapping mapping_;
+        fun_type fun_;
+
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:156]: I've gone with option (1). Since std::mdspan mappings are generally required to be copyable, adding an explicit std::is_copy_constructible_v<std::decay_t<Mapping>> constraint to the requires clause provides much cleaner error messages than letting the template instantiation fail deep in the internals. I have also updated the Doxygen documentation to reflect this requirement.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+                });
+        }
+        else
+        {
+            using index_type = typename Mapping::index_type;
+            index_type const bound =
+                static_cast<index_type>(m.extents().extent(RankIdx));
+            for (index_type i = index_type{0}; i < bound; ++i)
+            {
+                for_each_index_seq_impl<RankIdx + 1, ExPolicy>(m, f, idx..., i);
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: I've switched the internal counters to std::size_t and only cast to index_type at the very end when calling the user's function. Much safer!

**Comment by hkaiser** [comment]
> > Also, inpect reports an issue:
> > ```
> > /libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:
> >     Non-ASCII: (line 175)
> > ```
> 
> Fixed. I caught some en-dashes (–) in the comments that the inspect tool flagged. I've scrubbed the file and replaced them with standard ASCII hyphens.

Yep, LLMs do that to you ;-)

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 6 above)
Comment (hkaiser) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:350]: Does that mean that we could deduce the required iteration order from the `Mapping` that was passed in?

Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:350]: Yes, exactly!  I've just updated the parallel path to be layout-aware  now we check if the mapping exposes a stride() method at compile time. If it does, and we see stride(0) == 1 with a larger stride on the last dimension, we know it's column-major (layout_left). When that happens, it automatically flips the iteration: it parallelizes over the last dimension and sequentially recurses down to dimension 0. This guarantees we're always hitting contiguous memory and avoiding cache thrashing, regardless of whether they pass row-major or column-major mappings. Plus, since the detection is if constexpr, there's zero runtime overhead if the mapping doesn't even have a stride method.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+    {
+    private:
+        // ------------------------------------------------------------------ //
+        // Sequential overload (no execution policy) - constexpr, returns void.
+        // ------------------------------------------------------------------ //
+        template <typename Mapping, typename Fun>
+        // clang-format off
+        requires (
+            is_layout_mapping_v<Mapping> &&
+            std::is_copy_constructible_v<std::decay_t<Mapping>> &&
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Fixed..

File: libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+    {
+        return Rank;
+    }
+
+    [[nodiscard]] constexpr extents_type const& extents() const noexcept
+    {
+        return ext_;
+    }
+
+    extents_type ext_{};
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/tests/unit/algorithms/for_each_index.cpp]: fixed..

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 17]
```diff
@@ -0,0 +1,531 @@
...
+    // Sequential recursive core.
+    //
+    // Accumulates one index per dimension each level of recursion, then calls
+    // fun(i0, i1, ..., i_{rank-1}) when all indices are collected.
+    //
+    // Template parameter RankIdx is the *next* dimension to iterate (0-based).
+    // When RankIdx == Mapping::rank() we have the full index tuple.
+    template <std::size_t RankIdx, typename ExPolicy, typename Mapping,
+        typename Fun, typename... Indices>
+    HPX_HOST_DEVICE HPX_FORCEINLINE constexpr void for_each_index_seq_impl(
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Just for consistency, let's add `_right` to the names of this parallelization mode.

**Review by hkaiser — COMMENTED** [review]
Could you please also add tests for the version that assumes a column-major layout.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp (hunk 17 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Done! I've renamed the functions to _right for consistency, and also added full test coverage (both sequential and parallel) for the column-major layout_left path using a new mapping stub. Just pushed the update.

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 18]
```diff
@@ -0,0 +1,536 @@
...
+        requires (
+            is_layout_mapping_v<Mapping> &&
+            std::is_copy_constructible_v<std::decay_t<Fun>>
+        )
+        // clang-format on
+        friend constexpr void tag_fallback_invoke(
+            hpx::experimental::for_each_index_t, Mapping const& mapping,
+            Fun fun)
+        {
+            hpx::parallel::detail::for_each_index_seq_impl<0,
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp]: Fixed

File: libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp [hunk 19]
```diff
@@ -0,0 +1,536 @@
...
+
+        Mapping mapping_;
+        fun_type fun_;
+
+        template <typename M, typename F>
+        HPX_HOST_DEVICE for_each_index_iteration(M&& m, F&& f)
+          : mapping_(HPX_FORWARD(M, m))
+          , fun_(HPX_FORWARD(F, f))
+        {
+        }
```
[the remark this answers is not in the document]
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp:162]: Fixed

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
@arpittkhandelwal We're seeing compilation problems now: https://cdash.rostam.cct.lsu.edu/viewBuildError.php?buildid=48522. Could you have a look, please?

**Comment by arpittkhandelwal** [comment]
> @arpittkhandelwal We're seeing compilation problems now: https://cdash.rostam.cct.lsu.edu/viewBuildError.php?buildid=48522. Could you have a look, please?

Hi @hkaiser, yes I see the issue. The compiler is failing during the template instantiation of `HPX_INVOKE` because the `invoke` template alias hasn't been declared in that translation unit. We just need to add `#include <hpx/functional/invoke.hpp>` to the top of `libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp`.
Since this PR is already merged, should I create a new PR to push this fix?

**Comment by hkaiser** [comment]
> > @arpittkhandelwal We're seeing compilation problems now: https://cdash.rostam.cct.lsu.edu/viewBuildError.php?buildid=48522. Could you have a look, please?
> 
> Hi @hkaiser, yes I see the issue. The compiler is failing during the template instantiation of `HPX_INVOKE` because the `invoke` template alias hasn't been declared in that translation unit. We just need to add `#include <hpx/functional/invoke.hpp>` to the top of `libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp`. Since this PR is already merged, should I create a new PR to push this fix?

If that's the reason, then yes please cretae a new PR.

**Comment by arpittkhandelwal** [comment]
> > > @arpittkhandelwal We're seeing compilation problems now: https://cdash.rostam.cct.lsu.edu/viewBuildError.php?buildid=48522. Could you have a look, please?
> > 
> > 
> > Hi @hkaiser, yes I see the issue. The compiler is failing during the template instantiation of `HPX_INVOKE` because the `invoke` template alias hasn't been declared in that translation unit. We just need to add `#include <hpx/functional/invoke.hpp>` to the top of `libs/core/algorithms/include/hpx/parallel/algorithms/for_each_index.hpp`. Since this PR is already merged, should I create a new PR to push this fix?
> 
> If that's the reason, then yes please cretae a new PR.

 I have just created the new PR with the fix here: #7223.
